linera-storage/src/migration.rs

```diff
@@ -0,0 +1,296 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use linera_base::{
+    crypto::CryptoHash,
+    identifiers::{BlobId, ChainId, EventId},
+};
+use linera_views::{
+    batch::Batch,
+    store::{KeyValueDatabase, KeyValueStore, ReadableKeyValueStore, WritableKeyValueStore},
+    ViewError,
+};
+use serde::{Deserialize, Serialize};
+
+
+use crate::{
+    Clock,
+    db_storage::{DbStorage, DEFAULT_KEY, event_key, MultiPartitionBatch, ONE_KEY, RootKey},
+};
+
+#[derive(Default, Debug, Eq, PartialEq, Serialize, Deserialize)]
+#[repr(u8)]
+enum SchemaDescription {
+    /// Version 0, all the blobs, certificates, confirmed blocks, events and network description on the same partition.
+    /// This is marked as default since it does not exist in the old scheme and would be obtained from unwrap_or_default
+    #[default]
+    Version0,
+    /// Version 1, spreading by ChainId, CryptoHash, and BlobId.
+    Version1,
+}
+
+/// The key containing the schema of the storage.
+const SCHEMA_ROOT_KEY: &[u8] = &[0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233];
+
+#[derive(Debug, Serialize, Deserialize)]
+enum BaseKey {
+    ChainState(ChainId),
+    Certificate(CryptoHash),
+    ConfirmedBlock(CryptoHash),
+    Blob(BlobId),
+    BlobState(BlobId),
+    Event(EventId),
+    BlockExporterState(u32),
+    NetworkDescription,
+}
+
+const BASE_KEY_BLOCK_EXPORTER: u8 = 6;
+
+const UNUSED_EMPTY_KEY: &[u8] = &[];
+// The keys corresponding to `ChainState` are not moved.
+// The keys in `BlockExporterState` are moved, but belong to separate root keys.
+const MOVABLE_KEYS_0_1: &[u8] = &[1, 2, 3, 4, 5, 7];
+
+/// The total number of keys being migrated in a block.
+/// we use chunks to avoid OOM
+const BLOCK_KEY_SIZE: usize = 90;
+
+fn map_base_key(base_key: &[u8]) -> Result<(Vec<u8>, Vec<u8>), ViewError> {
```

Comment:
So given a base key, this returns the root (aka partition) key and a key for the specific entry inside that partition? (Does the latter have a name?) Would be good to document that.